src/main/java/tigase/xmpp/impl/BindResource.java

```diff
@@ -74,15 +83,21 @@ public String id() {
 		return ID;
 	}
 
-	public void setResourcePrefix(String resourcePrefix) {
-		this.resourcePrefix = resourcePrefix;
-		this.resourceDefPrefix =
-				Math.abs(DNSResolverFactory.getInstance().getDefaultHost().hashCode()) + "-" + (this.resourcePrefix != null ? this.resourcePrefix : resourceDefPrefix);
-	}
-
 	@Override
 	public boolean preProcess(Packet packet, XMPPResourceConnection session, NonAuthUserRepository repo,
 							  Queue<Packet> results, Map<String, Object> settings) {
+		if (packet.getServerAuthorisedStanzaFrom().isPresent()) {
+			// Correct user JID is already available in the packet so let's use it and short-circuit further processing.
+			packet.initVars(packet.getServerAuthorisedStanzaFrom().get(), packet.getStanzaTo());
```

Comment:
Again, we could use Optional.ifPresent() to pass function to call initVars(), but it is OK as we are leaving this method if we detect we can set it.